### PR TITLE
Servlets: Page Link Hit Counter

### PR DESCRIPTION
**Pull request by kristiyanpetkov**
In my PR i made a servlet which counts how many times a given link was hit.
Page first load:
screenshot from 2016-05-16 15-36-39

After few link hits:
screenshot from 2016-05-16 15-36-54

### CODE REVIEW

File: task1/src/main/java/com.clouway/LinkHitCounter.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+    String paramValue  = req.getParameter("link");
+
+    Integer accessCount = (Integer) session.getAttribute(paramValue);
+
+    if(paramValue==null){
+      session.setAttribute("first",0);
+      session.setAttribute("second",0);
+      session.setAttribute("third",0);
+    }
```
Comment (EmilGeorgiev) [inline task1/src/main/java/com.clouway/LinkHitCounter.java]: use map instead separate variable for each value. By this way we will avoid and many expectations in tests

File: task1/src/main/java/com.clouway/LinkHitCounter.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+    if (accessCount == null) {
+      accessCount = 0;
+    } else {
+      accessCount += 1;
+    }
+
+    if (paramValue != null) {
+      session.setAttribute(paramValue, accessCount);
+    }
```
Comment (EmilGeorgiev) [inline task1/src/main/java/com.clouway/LinkHitCounter.java]: this will be removed

File: task1/src/main/java/com.clouway/LinkHitCounter.java (hunk 1 above)
Comment (kristiyanpetkov) [inline task1/src/main/java/com.clouway/LinkHitCounter.java]: Fixed.

File: task1/src/main/java/com.clouway/LinkHitCounter.java (hunk 2 above)
Comment (kristiyanpetkov) [inline task1/src/main/java/com.clouway/LinkHitCounter.java]: Removed.

File: task1/src/main/java/com.clouway/LinkHitCounter.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+  @Override
+  protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
+
+    HttpSession session = req.getSession();
+
+    String paramValue = req.getParameter("link");
+
+    Map<String, Integer> accessCount = (Map<String, Integer>) session.getAttribute("links");
+    if (paramValue == null) {
+      accessCount = new HashMap<String, Integer>();
```
Comment (mgenov) [inline task1/src/main/java/com.clouway/LinkHitCounter.java]: You can do counting without the hardcoded "first", "second", "third.

File: task1/src/main/test/com.clouway/LinkHitCoutnerTest.java [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+    String expected = outputStream.toString();
+
+    assertThat(expected, containsString("<!DOCTYPE html>"));
+    assertThat(expected, containsString("<html>"));
+    assertThat(expected, containsString("<head><meta http-equiv='Content-Type' content='text/html; charset=UTF-8'>"));
+    assertThat(expected, containsString("<title>Session Test Servlet</title></head><body>"));
+    assertThat(expected, containsString("<p><a  href=\"linkhitcounter?link=first\">ABV</a>"));
+    assertThat(expected, containsString("<h2 style=\"color:blue;\">You have accessed this link 0 times.</h2>"));
+    assertThat(expected, containsString("<p><a  href=\"linkhitcounter?link=second\">GMAIL</a>"));
```
Comment (mgenov) [inline task1/src/main/test/com.clouway/LinkHitCoutnerTest.java]: Doesn't look like a correct way for testing counters. When HTML content is changed then and test need to be changed which is not a good symptom.

File: task1/target/task1/index.jsp [hunk 5]
```diff
@@ -0,0 +1,5 @@
+<html>
+<body>
+<h2>Hello World!</h2>
```
Comment (mgenov) [inline task1/target/task1/index.jsp]: JSP's are not required.

File: task1/pom.xml [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+      <scope>test</scope>
+    </dependency>
+
+    <dependency>
+      <artifactId>javax.servlet-api</artifactId>
+      <groupId>javax.servlet</groupId>
+      <version>3.1.0</version>
+    </dependency>
+
+    <dependency>
```
Comment (mgenov) [inline task1/pom.xml]: Jetty is included but I don't see any code related to it.

File: task1/src/main/java/com.clouway/LinkHitCounter.java (hunk 3 above)
Comment (kristiyanpetkov) [inline task1/src/main/java/com.clouway/LinkHitCounter.java]: Hardcoding removed.

File: task1/src/main/test/com.clouway/LinkHitCoutnerTest.java (hunk 4 above)
Comment (kristiyanpetkov) [inline task1/src/main/test/com.clouway/LinkHitCoutnerTest.java]: Fixed. Now i assert only the string message, not the whole html.

File: task1/target/task1/index.jsp (hunk 5 above)
Comment (kristiyanpetkov) [inline task1/target/task1/index.jsp]: Deleted. Actually it was automatically build, when creating the project.

File: task1/pom.xml (hunk 6 above)
Comment (kristiyanpetkov) [inline task1/pom.xml]: Fixed. Unused dependencies removed.

File: task1/src/main/java/com.clouway/LinkHitCounter.java [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+  }
+
+
+  public void printPage(PrintWriter out, Map<String, Integer> visitedPages) {
+    out.println("<!DOCTYPE html>");
+    out.println("<html>");
+    out.println("<head><meta http-equiv='Content-Type' content='text/html; charset=UTF-8'>");
+    out.println("<title>Session Test Servlet</title></head><body>");
+    out.println("<p><a  href=\"linkhitcounter?link=first\">ABV</a>");
+    if (visitedPages.containsKey("first")) {
```
Comment (mgenov) [inline task1/src/main/java/com.clouway/LinkHitCounter.java]: Adding of new page is forcing change here which is a SMELL.

File: task1/src/main/test/com.clouway/LinkHitCoutnerTest.java [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.hamcrest.CoreMatchers.containsString;
+import static org.hamcrest.MatcherAssert.assertThat;
+
+/**
+ * Created by clouway on 16.05.16.
+ */
+public class LinkHitCoutnerTest {
```
Comment (mgenov) [inline task1/src/main/test/com.clouway/LinkHitCoutnerTest.java]: There is a typo.

File: task1/src/main/test/com.clouway/LinkHitCoutnerTest.java [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+    assertThat(expected, containsString("You have accessed this link 1 times."));
+
+  }
+
+  @Test
+  public void visitSecondLinkTwice() throws Exception {
+    LinkHitCounter linkHitCounter = new LinkHitCounter();
+    final Map<String, Integer> visitedPages = new HashMap<String, Integer>();
+
+    context.checking(new Expectations() {{
```
Comment (mgenov) [inline task1/src/main/test/com.clouway/LinkHitCoutnerTest.java]: Looks like a lot of expectations. 

http://docs.oracle.com/javaee/6/api/javax/servlet/ServletRequest.html#getParameterMap()

File: task1/src/main/java/com.clouway/LinkHitCounter.java (hunk 7 above)
Comment (kristiyanpetkov) [inline task1/src/main/java/com.clouway/LinkHitCounter.java]: Fixed.

File: task1/src/main/test/com.clouway/LinkHitCoutnerTest.java (hunk 8 above)
Comment (kristiyanpetkov) [inline task1/src/main/test/com.clouway/LinkHitCoutnerTest.java]: Fixed.

File: task1/src/main/java/com.clouway/LinkHitCounter.java [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Created by clouway on 13.05.16.
+ */
+
+
```
Comment (panayotkulchev) [inline task1/src/main/java/com.clouway/LinkHitCounter.java]: /**
 * @author Panayot Kulchev <panayot.kulchev@clouway.com>
 */
And two lines to remove.

File: task1/src/main/java/com.clouway/LinkHitCounter.java [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+    if (visitedPages == null) {
+      visitedPages = new HashMap<String, Integer>();
+      session.setAttribute("links", visitedPages);
+      printPage(out, visitedPages);
+      return;
+    }
+
+    if (visitedPages.containsKey(paramValue)) {
+      Integer num = visitedPages.get(paramValue);
```
Comment (panayotkulchev) [inline task1/src/main/java/com.clouway/LinkHitCounter.java]: In the team usually when name variable that contain any number is called `count` or can leave it `nubmer` but do not use short

File: task1/src/main/java/com.clouway/LinkHitCounter.java [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+      printPage(out, visitedPages);
+    } else {
+      visitedPages.put(paramValue, 1);
+      session.setAttribute("links", visitedPages);
+      printPage(out, visitedPages);
+    }
+  }
+
+
+  public void printPage(PrintWriter out, Map<String, Integer> visitedPages) {
```
Comment (panayotkulchev) [inline task1/src/main/java/com.clouway/LinkHitCounter.java]: if you use this method only here make it `private`

File: task1/src/main/test/com.clouway/LinkHitCounterTest.java [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+
+  @After
+  public void cleanUp() throws IOException {
+    outputStream.close();
+  }
+
+
+  @Test
+  public void loadPage() throws Exception {
+    LinkHitCounter linkhitcounter = new LinkHitCounter();
```
Comment (panayotkulchev) [inline task1/src/main/test/com.clouway/LinkHitCounterTest.java]: Camel case is used for variables too in Java

File: task1/target/task1/META-INF/MANIFEST.MF [hunk 14]
```diff
@@ -0,0 +1,5 @@
+Manifest-Version: 1.0
```
Comment (panayotkulchev) [inline task1/target/task1/META-INF/MANIFEST.MF]: .gitIgnore

File: task1/task1.iml [hunk 15]
```diff
@@ -0,0 +1,41 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (panayotkulchev) [inline task1/task1.iml]: .gitIgnore

**Comment by panayotkulchev** [comment]
Reviewed by panayotkulchev

File: task1/src/main/test/com.clouway/LinkHitCoutnerTest.java (hunk 9 above)
Comment (kristiyanpetkov) [inline task1/src/main/test/com.clouway/LinkHitCoutnerTest.java]: I made tests with faked interfaces. No more expectations.

File: task1/src/main/java/com.clouway/LinkHitCounter.java (hunk 10 above)
Comment (kristiyanpetkov) [inline task1/src/main/java/com.clouway/LinkHitCounter.java]: Fixed.

File: task1/src/main/java/com.clouway/LinkHitCounter.java (hunk 11 above)
Comment (kristiyanpetkov) [inline task1/src/main/java/com.clouway/LinkHitCounter.java]: Fixed.

File: task1/src/main/java/com.clouway/LinkHitCounter.java (hunk 12 above)
Comment (kristiyanpetkov) [inline task1/src/main/java/com.clouway/LinkHitCounter.java]: Fixed.

File: task1/src/main/test/com.clouway/LinkHitCounterTest.java (hunk 13 above)
Comment (kristiyanpetkov) [inline task1/src/main/test/com.clouway/LinkHitCounterTest.java]: Fixed.

File: task1/target/task1/META-INF/MANIFEST.MF (hunk 14 above)
Comment (kristiyanpetkov) [inline task1/target/task1/META-INF/MANIFEST.MF]: Added.

File: task1/task1.iml (hunk 15 above)
Comment (kristiyanpetkov) [inline task1/task1.iml]: Added.

**Comment by kristiyanpetkov** [comment]
All comments fixed!

**Comment by mgenov** [comment]
LGTM !

**Pull request merged**